pkg/simple/client/monitoring/metricsserver/metricsserver.go

```diff
@@ -319,118 +520,212 @@ func (m metricsServer) GetNamedMetricsOverTime(metrics []string, start, end time
-		}
+	if !m.metricsAPIAvailable {
+		klog.Warningf("Metrics API not available.")
+		return m.parseErrorResp(metrics, errors.New("Metrics API not available."))
+	}
+
+	switch opts.Level {
+	case monitoring.LevelNode:
+		return m.GetNodeLevelNamedMetricsOverTime(metrics, start, end, step, opts)
+	case monitoring.LevelPod:
+		return m.GetPodLevelNamedMetricsOverTime(metrics, start, end, step, opts)
+	default:
+		return res
+	}
+
+}
+
+func (m metricsServer) GetNodeLevelNamedMetricsOverTime(metrics []string, start, end time.Time, step time.Duration, opts *monitoring.QueryOptions) []monitoring.Metric {
+	var res []monitoring.Metric
+	edgeNodes, err := m.listEdgeNodes()
+	if err != nil {
+		klog.Errorf("List edge nodes error %v\n", err)
+		return m.parseErrorResp(metrics, err)
+	}
+
+	edgeNodeNamesFiltered := m.filterEdgeNodeNames(edgeNodes, opts)
+	if len(edgeNodeNamesFiltered) == 0 {
+		klog.V(4).Infof("No edge node metrics is requested")
+		return res
+	}
+
+	metricsResult, err := m.getNodeMetricsFromMetricsAPI()
+	if err != nil {
+		klog.Errorf("Get edge node metrics error %v\n", err)
+		return m.parseErrorResp(metrics, err)
+	}
+
+	metricsMap := make(map[string]bool)
+	for _, m := range metrics {
+		metricsMap[m] = true
+	}
 
-		edgeNodeNamesFiltered := m.filterEdgeNodeNames(edgeNodes, opts)
-		if len(edgeNodeNamesFiltered) == 0 {
-			klog.V(4).Infof("No edge node metrics is requested")
-			return res
+	status := make(map[string]v1.NodeStatus)
+	for n, _ := range edgeNodeNamesFiltered {
+		status[n] = edgeNodes[n].Status
+	}
+
+	nodeMetrics := make(map[string]*monitoring.MetricData)
+	for _, enm := range edgeNodeMetrics {
+		_, ok := metricsMap[enm]
+		if ok {
+			nodeMetrics[enm] = &monitoring.MetricData{MetricType: monitoring.MetricTypeMatrix}
 		}
+	}
 
-		metricsResult, err := m.getNodeMetricsFromMetricsAPI()
-		if err != nil {
-			klog.Errorf("Get edge node metrics error %v\n", err)
-			return m.parseErrorResp(metrics, err)
+	var usage v1.ResourceList
+	var cap v1.ResourceList
+	for _, m := range metricsResult.Items {
+		_, ok := edgeNodeNamesFiltered[m.Name]
+		if !ok {
+			continue
 		}
 
-		metricsMap := make(map[string]bool)
-		for _, m := range metrics {
-			metricsMap[m] = true
+		m.Usage.DeepCopyInto(&usage)
+		status[m.Name].Capacity.DeepCopyInto(&cap)
+
+		metricValues := make(map[string]*monitoring.MetricValue)
+
+		for _, enm := range edgeNodeMetrics {
+			metricValues[enm] = &monitoring.MetricValue{
+				Metadata: make(map[string]string),
+			}
+			metricValues[enm].Metadata["node"] = m.Name
+			metricValues[enm].Metadata["role"] = "edge"
+		}
+		for _, addr := range status[m.Name].Addresses {
+			if addr.Type == v1.NodeInternalIP {
+				for _, enm := range edgeNodeMetrics {
+					metricValues[enm].Metadata["host_ip"] = addr.Address
+				}
+				break
+			}
 		}
 
-		status := make(map[string]v1.NodeStatus)
-		for n, _ := range edgeNodeNamesFiltered {
-			status[n] = edgeNodes[n].Status
+		_, ok = metricsMap[metricsNodeCPUUsage]
```

Comment:
DRY, I think we can refactor these code with for range, for example
```
for k, v := range metricsMap {
  switch k {
  case metricsNodeCPUUsage:
      .....
  case metricsNodeCPUTotal:
    ....
  }
}
```

Reply:
fixed

Reply:
`range` a `map` returns `key, value`, the second returned value is not `ok`,  checkout this example https://gobyexample.com/range

Reply:
The  type of  the values in the map here is bool .  Maybe `ok` in golang has special meaning.  I think I can modify it to improve readability.